File: Binary_Search/Binary_Search.py
# Function for Binary Search
def BinarySearch(arr, low, high, num):
    if(high >= low):
        middle = (high + low) // 2
        #if the num is less than the middle number
        if(num < arr[middle]):
            return BinarySearch(arr, low, middle-1, num) 
        #if the num is more than the middle number
        elif(num > arr[middle]):
            return BinarySearch(arr, middle+1, high, num) 
        else:
            return middle
    else:
        return -1

arr = []
n = int(input("Enter number of elements you want in the array: "))
print("Note: Enter " + str(n) + " numbers in SORTED ORDER")

for i in range(n):
    element = int(input("Enter element " + str(i+1) + " : "))
    arr.append(element)
    
desired = int(input("Enter Element you want to Search: "))
index = BinarySearch(arr, 0, len(arr)-1, desired)

if(index != -1):
    print(" ----Number Found at index: " + str(index+1) + " ----")
else:
    print("----Number Not Found! Try again----")



'''
Input:
num = 7
array = {1, 2, 3, 4, 5, 6, 7}
desired = 4

Output:
----Number Found at index: 4 ----

Input:
num = 5
array = {1, 3, 5, 7, 9}
desired = 2

Output:
----Number Not Found! Try again----
'''

File: Binary_Search/Binary_Search.dart
//Binary Search is an efficient algorithm for finding an item from a sorted list of items.
//This is the implementation of Binary Search in dart language to insert and search the elements.
import 'dart:io';

void binarySearch(List<int> list, int x){
    int size = list.length;
    int low = 0;
    int high = size - 1;
    while(low <= high){
        int mid = ((low + high) / 2).floor();
        if(list[mid] == x){
            print('Found $x in the List at Index $mid!');
            return;
        }
        if(list[mid] > x){
            high = mid - 1;
            continue;
        }
        low = mid + 1;
    }
    print('Did not $x in the List!');
    return;
}

main(){
    List<int> list = new List();
    int task = 0;
    print('Binary Search Program');
    while(task != 4){
        print('0 - Display List\n1 - Insert element in List \n2 - Search element in List\n3 - Quit\nSelect task:-');
        try{
            task = int.parse(stdin.readLineSync());
        } on Exception{
            print('Please enter valid Integer value!');
            continue;
        }
        switch(task){
            case 0:
                if(list.length == 0){
                    print('Empty List!');
                } 
                else{
                    print('List:-');
                    list.forEach((val){
                        print(val);
                    });
                }
                break;
            case 1:
                int val = null;
                print('Enter value that needs to be inserted in List:- ');
                try{
                    val = int.parse(stdin.readLineSync());
                } on Exception{
                    print('Please enter valid Integer value!');
                    continue;
                }
                if(val == null){
                    print('Failed to insert element in List!');
                    continue;
                }
                list.add(val);
                list.sort();
                print('Successfully inserted $val in List!');
                break;

            case 2:
                int val = null;
                print('Enter value that needs to be searched in List:- ');
                try{
                    val = int.parse(stdin.readLineSync());
                } on Exception{
                    print('Please enter valid Integer value!');
                    continue;
                }
                if(val == null){
                    print('Failed to search element in List!');
                    continue;
                }
                binarySearch(list, val);
                break;
            
            case 3:
                print('Binary Search Program Complete');
                break;
                default:
                continue;
        }
    }
}


/*
Sample Input/Output:
Binary Search Program
0 - Display List
1 - Insert element in List
2 - Search element in List
3 - Quit
Select task:-		//to add element
1
Enter value that needs to be inserted in List:- 
1
Successfully inserted 1 in List!
0 - Display List
1 - Insert element in List
2 - Search element in List
3 - Quit
Select task:-
1
Enter value that needs to be inserted in List:- 
2
Successfully inserted 2 in List!
0 - Display List
1 - Insert element in List
2 - Search element in List
3 - Quit
Select task:-
1
Enter value that needs to be inserted in List:- 
3
Successfully inserted 3 in List!
0 - Display List
1 - Insert element in List
2 - Search element in List
3 - Quit
Select task:-		//to display the list
0
List:-
1
2
3
0 - Display List
1 - Insert element in List
2 - Search element in List
3 - Quit
Select task:-		//to search the element
2
Enter value that needs to be searched in List:- 
2
Found 2 in the List at Index 1!
0 - Display List
1 - Insert element in List
2 - Search element in List
3 - Quit
Select task:-		//quiting the program
3
Binary Search Program Complete
*/
